### PR TITLE
add structN2V ability to method w membrane example

### PR DESCRIPTION
**Pull request by colemanbroad**

### CODE REVIEW

File: n2v/internals/N2V_DataWrapper.py [hunk 1]
```diff
@@ -1,7 +1,8 @@
 from keras.utils import Sequence
 
 import numpy as np
-
+# from ..utils import n2v_utils
```
Comment (tibuch) [inline n2v/internals/N2V_DataWrapper.py:4]: Please remove commented lines.

File: n2v/internals/N2V_DataWrapper.py [hunk 2]
```diff
@@ -34,6 +35,9 @@ def __init__(self, X, Y, batch_size, perc_pix=0.198, shape=(64, 64),
         self.range = np.array(self.X.shape[1:-1]) - np.array(self.shape)
         self.dims = len(shape)
         self.n_chan = X.shape[-1]
+        self.structN2Vmask = structN2Vmask
+        # self.structN2Vmask = np.array([[0,0,1,1,1,0,0]])
```
Comment (tibuch) [inline n2v/internals/N2V_DataWrapper.py:39]: Please remove commented lines.

File: n2v/internals/N2V_DataWrapper.py [hunk 3]
```diff
@@ -34,6 +35,9 @@ def __init__(self, X, Y, batch_size, perc_pix=0.198, shape=(64, 64),
...
+        print("StructN2V Mask is: ", self.structN2Vmask)
```
Comment (tibuch) [inline n2v/internals/N2V_DataWrapper.py:40]: I would suggest to only print the structN2Vmask if it is not None.

File: n2v/internals/N2V_DataWrapper.py [hunk 4]
```diff
@@ -81,9 +85,39 @@ def __getitem__(self, i):
...
+        ## combine all coords (ndim, npts,) with all displacements (ncoords,ndim,)
+        mix = (dx.T[...,None] + coords[None])
+        mix = mix.transpose([1,0,2]).reshape([ndim,-1]).T
+        ## stay within patch boundary
+        mix = mix.clip(min=np.zeros(ndim),max=np.array(patch.shape)-1).astype(np.uint)
+        ## replace neighbouring pixels with random values from flat dist
+        patch[tuple(mix.T)] = np.random.rand(mix.shape[0])*4 - 2
+        # patch[tuple(coords)] = 2
+
+
```
Comment (tibuch) [inline n2v/internals/N2V_DataWrapper.py:116]: Please remove empty code lines and commented code which is not used.

File: n2v/models/n2v_config.py [hunk 5]
```diff
@@ -223,6 +227,8 @@ def _is_int(v,low=None,high=None):
         )
         ok['n2v_manipulator']       = self.n2v_manipulator in ['normal_withoutCP', 'uniform_withCP', 'normal_additive',
                                                                'normal_fitted', 'identity']
+        # ok['structN2Vmask'] = self.structN2Vmask.ndim = self.ndim and all([x in [0,1] for x in self.structN2Vmask.flat])
+        ok['structN2Vmask'] = True
```
Comment (tibuch) [inline n2v/models/n2v_config.py:231]: This should test if the dimensions of the mask are odd and if all entries are either 0 or 1.

File: n2v/models/n2v_standard.py [hunk 6]
```diff
@@ -63,10 +63,11 @@ class N2V(CARE):
 
     def __init__(self, config, name=None, basedir='.'):
         """See class docstring."""
+        # self._config_class
```
Comment (tibuch) [inline n2v/models/n2v_standard.py:66]: Please remove this line.

File: n2v/models/n2v_standard.py [hunk 7]
```diff
@@ -88,6 +89,11 @@ def __init__(self, config, name=None, basedir='.'):
         self.keras_model = self._build()
         if config is None:
             self._find_and_load_weights()
+        
+        
+        print("config.structN2Vmask", config.structN2Vmask)
```
Comment (tibuch) [inline n2v/models/n2v_standard.py:94]: Does this print statement serve a debugging purpose? Would it be possible to remove it or is it necessary?

File: n2v/models/n2v_standard.py [hunk 8]
```diff
@@ -209,7 +215,7 @@ def train(self, X, validation_X, epochs=None, steps_per_epoch=None):
         # a masking channel. The N2V_DataWrapper will take care of the pixel masking and manipulating.
         training_data = N2V_DataWrapper(X, np.concatenate((X, np.zeros(X.shape, dtype=X.dtype)), axis=axes.index('C')),
                                                     self.config.train_batch_size, self.config.n2v_perc_pix,
-                                                    self.config.n2v_patch_shape, manipulator)
+                                                    self.config.n2v_patch_shape, manipulator, structN2Vmask=np.array(self.config.structN2Vmask))
```
Comment (tibuch) [inline n2v/models/n2v_standard.py:218]: If the `structN2Vmask` is `None` this will be converted to a `array(None, dtype=object)` object. Which is not `None` and therefore the N2V_DataWrapper will try to use the StructN2V code, which will fail in the normal N2V case.

File: n2v/utils/n2v_utils.py [hunk 9]
```diff
@@ -73,7 +73,6 @@ def local_gaussian(patch, coords, dims):
         return vals
     return local_gaussian
 
-
```
Comment (tibuch) [inline n2v/utils/n2v_utils.py:76]: According to [PEP8 Python formatting guide lines](https://www.python.org/dev/peps/pep-0008/#blank-lines) top-level functions should be surrounded by two blank lines.

File: README.md [hunk 10]
```diff
@@ -67,3 +68,6 @@ Have a look at our jupyter notebook:
   year={2019}
 }
 ```
+
+see [here](https://github.com/mpicbg-csbd/structured_N2V_paper) for citing `StructN2V`.
```
Comment (tibuch) [inline README.md:72]: This link returns a 404 for me.

File: n2v/internals/N2V_DataWrapper.py (hunk 4 above)
Comment (tibuch) [reply n2v/internals/N2V_DataWrapper.py:113]: Please :pray: ? :)

File: README.md (hunk 10 above)
Comment (colemanbroad) [reply README.md:72]: hmmm not for me. wonder what's going on.

Comment (colemanbroad) [reply README.md:72]: the _upper_ link `https://github.com/juglab/n2v/tree/structn2v/examples/2D/structN2V_2D_synth_mem` gives 404, but i guess that's because the we haven't merged yet ;)

File: n2v/internals/N2V_DataWrapper.py (hunk 4 above)
Comment (colemanbroad) [reply n2v/internals/N2V_DataWrapper.py]: ok fixed the remaining whitespace :)

File: README.md (hunk 10 above)
Comment (tibuch) [reply README.md:72]: hmm. this one works for you? https://github.com/mpicbg-csbd/structured_N2V_paper

**Comment by colemanbroad** [comment]
oops! needed to make the repo public :wink:
only worked for me as i was already logged in.
Should be public anyways. Happy to do that.
